File: python/sdk/src/windy_drops/lib/skill_md.py
"""skill_md.py — parse a SKILL.md file into YAML frontmatter dict + body.

Mirrors packages/sdk/src/lib/skill-md.ts (TS sibling). Both parsers MUST
extract the same frontmatter object for the same input file.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import yaml

FRONTMATTER_RE = re.compile(
    r"^---\r?\n(?P<yaml>[\s\S]*?)\r?\n---\r?\n?(?P<body>[\s\S]*)$"
)


class SkillMdError(Exception):
    """Raised when SKILL.md is missing or malformed."""

    def __init__(self, message: str, file_path: Path | str) -> None:
        super().__init__(f"{file_path}: {message}")
        self.file_path = str(file_path)


@dataclass(frozen=True)
class ParsedSkillMd:
    file_path: Path
    frontmatter: dict[str, Any]
    body: str
    frontmatter_start_line: int
    body_start_line: int


def read_skill_md(drop_dir: str | Path) -> ParsedSkillMd:
    file_path = Path(drop_dir).resolve() / "SKILL.md"
    try:
        raw = file_path.read_text(encoding="utf-8")
    except FileNotFoundError as e:
        raise SkillMdError(f"SKILL.md not found ({e})", file_path) from e

    match = FRONTMATTER_RE.match(raw)
    if not match:
        raise SkillMdError(
            "no YAML frontmatter (must begin with `---` and contain a closing `---`)",
            file_path,
        )

    try:
        frontmatter = yaml.safe_load(match.group("yaml") or "")
    except yaml.YAMLError as e:
        raise SkillMdError(f"YAML parse error: {e}", file_path) from e

    if not isinstance(frontmatter, dict):
        raise SkillMdError("frontmatter must be a YAML object", file_path)

    yaml_line_count = (match.group("yaml") or "").count("\n") + 1
    body_start_line = 2 + yaml_line_count

    return ParsedSkillMd(
        file_path=file_path,
        frontmatter=frontmatter,
        body=match.group("body") or "",
        frontmatter_start_line=1,
        body_start_line=body_start_line,
    )

File: packages/sdk/src/lib/skill-md.ts
// skill-md.ts — parse a SKILL.md file into its YAML frontmatter object + body.
// Used by every SDK command that needs to read or rewrite a drop's manifest.

import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { parse as parseYaml } from "yaml";

export interface ParsedSkillMd {
  filePath: string;
  frontmatter: Record<string, unknown>;
  body: string;
  // Line at which the frontmatter starts in the source (always 1 for valid files).
  frontmatterStartLine: number;
  // Line at which the body starts (after the closing ---).
  bodyStartLine: number;
}

export class SkillMdError extends Error {
  constructor(message: string, readonly filePath: string) {
    super(`${filePath}: ${message}`);
    this.name = "SkillMdError";
  }
}

const FRONTMATTER_RE = /^---\r?\n([\s\S]*?)\r?\n---\r?\n?([\s\S]*)$/;

export function readSkillMd(dropDir: string): ParsedSkillMd {
  const filePath = resolve(dropDir, "SKILL.md");
  let raw: string;
  try {
    raw = readFileSync(filePath, "utf-8");
  } catch (e) {
    throw new SkillMdError(`SKILL.md not found (${(e as Error).message})`, filePath);
  }

  const match = raw.match(FRONTMATTER_RE);
  if (!match) {
    throw new SkillMdError(
      "no YAML frontmatter (must begin with `---` and contain a closing `---`)",
      filePath,
    );
  }

  let frontmatter: unknown;
  try {
    frontmatter = parseYaml(match[1] ?? "");
  } catch (e) {
    throw new SkillMdError(`YAML parse error: ${(e as Error).message}`, filePath);
  }

  if (typeof frontmatter !== "object" || frontmatter === null || Array.isArray(frontmatter)) {
    throw new SkillMdError("frontmatter must be a YAML object", filePath);
  }

  const body = match[2] ?? "";
  const yamlLineCount = (match[1] ?? "").split(/\r?\n/).length;
  // 1: opening `---`; +yamlLineCount; +1: closing `---`. Body starts on the next line.
  const bodyStartLine = 2 + yamlLineCount;

  return {
    filePath,
    frontmatter: frontmatter as Record<string, unknown>,
    body,
    frontmatterStartLine: 1,
    bodyStartLine,
  };
}
